aiida/tools/graph/graph_traversers.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Module for functions to traverse AiiDA graphs."""
+
+from __future__ import absolute_import
+from numpy import inf
+
+
+def traverse_graph(starting_pks, max_iterations=None, get_links=False, **kwargs):
+    """
+    This function will return the set of all nodes that can be connected
+    to a list of initial nodes through any sequence of specified authorized
+    links and directions.
+
+    :type starting_pks:  list
+    :param starting_pks:
+        A list with the (valid) pks of all starting nodes.
+
+    :type max_iterations: int or None
+    :param max_iterations:
+        The number of iterations to apply the set of rules (a value of 'None' will
+        iterate until no new nodes are added).
+
+    :type get_links:  bool
+    :param get_links:
+        Pass True to also return the links between all nodes (found + initial).
+
+    :type traverse_rules: aiida.common.links.GraphTraversalRule
```

Comment:
Replace this by explicitly excepted keywords as in export util now: e.g.
```
:param input_calc_forward: Follow INPUT_CALC links in the forward direction (recursively).
```
and add a `.. note:` above explaining that some of these will not be allowed to be passed depending on the operation defined by the `graph_traversal_rules`